server/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java

```diff
@@ -152,6 +155,13 @@ static boolean isValidAffixKey(String key) {
     public synchronized Settings validateUpdate(Settings settings) {
         final Settings current = Settings.builder().put(this.settings).put(settings).build();
         final Settings previous = Settings.builder().put(this.settings).put(this.lastSettingsApplied).build();
+
+        if (INDEX_BALANCE_FACTOR_SETTING.get(current) + SHARD_BALANCE_FACTOR_SETTING.get(current) == 0.0f) {
+            throw new IllegalArgumentException(
+                "[" + INDEX_BALANCE_FACTOR_SETTING.getKey() + "] and [" + SHARD_BALANCE_FACTOR_SETTING.getKey() +
+                "] cannot both be set to 0.0");
+        }
```

Comment:
Hm, this is an abstract class which, until now, knows nothing about any concrete setting, it works purely on semantics. Can't we do this validation somewhere else?

Reply:
yep, you're right, how about moving them to ClusterSettings (check new commit).